### ISSUE TITLE
KTransformers

### ISSUE THREAD

**Issue by ErvinXie** [pytorch-fdn/ecosystem#57]
### Contact emails

james0zan@gmail.com, xwy21@mails.tsinghua.edu.cn

### Project summary

Efficient inference and fine-tuning of large language models through CPU-GPU heterogeneous computing.

### Project description

KTransformers is a backend library that focuses on high-performance kernels for CPUs across diverse architectures and instruction sets, while leveraging existing GPU kernels. Through fine-grained module partitioning and hybrid computing, it enables resource-constrained users to efficiently deploy and fine-tune extremely large MoE models locally, e.g., deploying a trillion-parameter MoE model with a single consumer-grade RTX 4090 GPU, and fine-tuning it with only two RTX 4090s.

It provides two primary modules:

- **kt-kernel**: high-performance LLM operators such as fused MoE and sparse attention, with optimized CPU backends for heterogeneous deployments.
- **kt-sft**: fine-tuning integration built on top of popular training stacks, enabling resource-efficient training for large models.

The project also offers an easy-to-use CLI interface, and integrates smoothly with PyTorch-based engines such as HF-Transformers and SGLang as a backend acceleration library, enabling GPU–CPU heterogeneous computing. By combining KTransformers with HF-Transformers, users can achieve 4–20x local inference speedups.

### Are there any other projects in the PyTorch Ecosystem similar to yours?  If, yes, what are they?

No. While engines such as HF-Transformers provide LLM fine-tuning and deployment capabilities, they primarily focus on homogeneous setups in resource-constrained scenarios. KTransformers is collaborative and complementary with them: it adds CPU–GPU heterogeneous acceleration, enabling flexible module offloading to CPUs while still achieving strong performance.

### Project repo URL

https://github.com/kvcache-ai/ktransformers

### Additional repos in scope of the application

No response

### Project license

Apache-2.0

### GitHub handles of the project maintainer(s)

[[@ErvinXie](https://github.com/ErvinXie)](https://github.com/ErvinXie), [[@chenht2022](https://github.com/chenht2022)](https://github.com/chenht2022), [[@ovowei](https://github.com/ovowei)](https://github.com/ovowei), [[@KMSorSMS](https://github.com/KMSorSMS)](https://github.com/KMSorSMS), [[@ouqingliang](https://github.com/ouqingliang)](https://github.com/ouqingliang), [[@SkqLiao](https://github.com/SkqLiao)](https://github.com/SkqLiao), [[@JimmyPeilinLi](https://github.com/JimmyPeilinLi)](https://github.com/JimmyPeilinLi), [[@mrhaoxx](https://github.com/mrhaoxx)](https://github.com/mrhaoxx), [[@Atream](https://github.com/Atream)](https://github.com/Atream), [[@Azure-Tang](https://github.com/Azure-Tang)](https://github.com/Azure-Tang), [[@qiyuxinlin](https://github.com/qiyuxinlin)](https://github.com/qiyuxinlin)

### Is there a corporate or academic entity backing this project?  If so, please provide the name and URL of the entity.

[MADSys Lab](https://madsys.cs.tsinghua.edu.cn/) @ Tsinghua University and [Approaching.AI](http://approaching.ai/).

### Website URL

https://kvcache-ai.github.io/ktransformers/

### Documentation

https://kvcache-ai.github.io/ktransformers/index.html

### How do you build and test the project today (continuous integration)?  Please describe.

Our CI pipeline ensures:

- Documentation build and deploy on GitHub Actions across Ubuntu/macOS/Windows for PRs and main.
- PR kt-kernel CPU tests on a self-hosted runner, gated by a `run-ci` label and non-draft status.
- Docker image build and push on release or manual trigger via GitHub Actions.
- PyPI wheel build and publish triggered by `version.py` updates on main (with GPU self-hosted builders for wheel builds).

### Version of PyTorch

2.9.0, 2.9.1, 2.10.0

### Components of PyTorch

- PyTorch C++/CUDA extensions and runtime integration
- Standard PyTorch Python tensor APIs

### How long do you expect to maintain the project?

We are committed to maintaining and evolving the project for the foreseeable future.

### Additional information

KTransformers is a research project conducted at MADSys Lab @ Tsinghua University under the supervision of Associate Professor Mingxing Zhang. A paper based on this project was accepted at SOSP 2025: [[SOSP 2025 Paper](https://dl.acm.org/doi/10.1145/3731569.3764843)](https://dl.acm.org/doi/10.1145/3731569.3764843)

KTransformers is actively developed with regular updates, documentation, and model support. It currently has 93 contributors and 11 maintainers from diverse organizations including Intel, Huawei, and MooreThreads, etc.

**Comment by jspisak** [comment]
Thanks @ErvinXie for submitting this! I just saw a few things called out below. Can you take a look?

Requirement | Status
-- | --
PyTorch Support | Indirect
Working CI |  ✅
Governance | ✅
User Experience | ✅
Quick Start | ✅
Documentation | ✅
Permissive License | ✅
Functional Testing | ❓ Unknown
Packages | ✅
Ongoing Maintenance | ✅
Core Maintainers | ✅
Active Contributors | ✅
90d Commits | ✅
≥200 Stars | ✅
LICENSE | ✅
README | ✅
CONTRIBUTING.md | ✅
CODEOWNERS/MAINTAINERS | ✅
CODE_OF_CONDUCT | ✅

**Comment by ErvinXie** [comment]
Hi @jspisak , thanks for the review! Here are some clarifications on the two flagged items:

  PyTorch Support                                                 

I understand the "Indirect" classification — let me provide more context. KTransformers has recently been refactored into a modular architecture:

- kt-kernel (the main component): Provides CPU-optimized kernels (AMX/AVX512) for MoE expert computation. It declares torch>=2.0.0 as a hard dependency and uses torch.Tensor as the native data structure for all inputs/outputs. The C++ bindings (via pybind11) create and operate on torch tensors directly. It integrates with SGLang as the serving framework, providing CPU expert execution, while SGLang handles GPU inference — both built on PyTorch.
- kt-sft: The fine-tuning component, which heavily uses PyTorch — all operators inherit from torch.nn.Module, training uses torch.utils.data.DataLoader, and it builds C++/CUDA extensions via torch.utils.cpp_extension.CUDAExtension.

So while kt-kernel is not building nn.Module subclasses itself, PyTorch is a required dependency and torch.Tensor is the fundamental data type throughout the project. The relationship is perhaps best described as: kt-kernel extends PyTorch's compute capabilities by providing optimized CPU backends for heterogeneous inference, and integrates with PyTorch-based engines (SGLang, HF-Transformers) as a backend acceleration library.

Functional Testing

We have functional tests at multiple levels:

- Automated CI: https://github.com/kvcache-ai/ktransformers/blob/main/.github/workflows/kt-kernel-tests.yml runs per-commit tests on every PR to main/develop on a self-hosted CPU runner.
- Accuracy tests: Tests in kt-kernel/test/per_commit/ (e.g., test_moe_amx_accuracy_int4.py) compare kernel outputs against PyTorch reference implementations with numerical tolerance checks.
- Comprehensive kernel examples/tests: The https://github.com/kvcache-ai/ktransformers/tree/main/kt-kernel/examples directory contains 30+ functional test scripts covering attention, RoPE, MLA (Multi-Head Latent Attention), linear layers, MLP, gating, and MoE kernels across multiple precision formats (FP8, BF16, AWQ 4-bit, INT8). It also includes PyTorch reference implementations (e.g., torch_attention.py, test_mla_torch.py) for correctness validation, and end-to-end model tests for DeepSeek V3 inference and prefill.
- C++ unit tests: Low-level kernel tests in kt-kernel/operators/.
- Public benchmarks: We maintain a public benchmark leaderboard at https://ktransformers.net/en/benchmarks with reproducible performance results across multiple models (DeepSeek-V3.2, Qwen3.5 series, GLM-5 744B, etc.) on standardized hardware (RTX 5090 + AMD EPYC 9355), reporting decode TPS with 5-7 test iterations per configuration.

Let me know if you need further details!

**Comment by jspisak** [comment]
thanks for clarifying - super helpful!  Based on this @michellearoth - i am good to proceed..

**Comment by michellearoth** [comment]
Approved! 

<img width="780" height="826" alt="Image" src="https://github.com/user-attachments/assets/c6bf17a9-66f3-4815-9b1e-d71f461f4a17" />
